README.org
#+LATEX_HEADER: \usepackage[margin=1.5cm,includefoot]{geometry}
#+LATEX_HEADER: \usepackage[none]{hyphenat}
#+LATEX_HEADER: \usepackage{array}
#+LATEX_HEADER: \newcolumntype{$}{>{\global\let\currentrowstyle\relax}}
#+LATEX_HEADER: \newcolumntype{^}{>{\currentrowstyle}}
#+LATEX_HEADER: \newcommand{\rowstyle}[1]{\gdef\currentrowstyle{#1} #1\ignorespaces}
#+LATEX_HEADER: \usepackage{indentfirst}

#+OPTIONS: toc:nil title:nil

\begin{titlepage}
  \begin{center}
    \line(1,0){300} \\
    [5mm]
    \huge{\bfseries Differential Equations Coursework} \\
    \huge{\bfseries Aeroplane Landing Modelling} \\
    [5mm]
    \huge{Gleb Dianov} \\
  \end{center}
\end{titlepage}

\newpage

\tableofcontents

\newpage

* Simplifying situation and setting up the model
** Given data

I've been given data about a landing aeroplane after touchdown. The mass of the aeroplane is $120 000$ kg. Initially air resistance slows the aeroplane and then, when it is slow enough, wheel brakes add a constant force to fully stop the aeroplane. The table below shows the aeroplane's speed, $v$ ms^{-1}, $t$ seconds after touchdown.

#+ATTR_LATEX: :align |rl|rl|
|----+----+----+----|
|  t |  v |  t |  v |
|----+----+----+----|
|  0 | 96 | 14 | 34 |
|  1 | 89 | 15 | 31 |
|  2 | 82 | 16 | 27 |
|  3 | 77 | 17 | 24 |
|  4 | 72 | 18 | 21 |
|  5 | 68 | 19 | 18 |
|  6 | 64 | 20 | 16 |
|  7 | 61 | 21 | 13 |
|  8 | 58 | 22 | 10 |
|  9 | 55 | 23 |  8 |
| 10 | 50 | 24 |  5 |
| 11 | 46 | 25 |  3 |
| 12 | 41 | 26 |  0 |
| 13 | 38 |    |    |
|----+----+----+----|

The image below shows a chart of the original data.

#+NAME: original_graph
[[./original_graph.png]]

** Modeling assumptions

- *The aeroplane can be modeled as a point particle*. This assumption allows us to negate any complex resistances that would occur. Modeling this way is appropriate because we are not given velocity (only speed) of the aeroplane at different points in time, so we are not interested in any rotations of the plane.
- *The runway can be modeled as a plane*. Without this assumption we would need to consider the possibility that the runway has a complex shape, for example it can have hills and/or pits.
- *The runway is horizontal*. Without this assumption we would need to include components of the weight and the reaction force in the model which would increase the complexity of the model and affect the calculations.
- *The aeroplane is moving in the horizontal plane*. This assumption allows us to negate any vertical forces.
- *The constant force from the wheel breaks (after the breaks are applied) and the air resistance are the only forces acting on the aeroplane in the horizontal plane*. This assumption allows us to say that any other resistance is negligible and there are no other forces, alternatively we can say that the air resistance model is a model of the total resistance force.
- *The given values for speed are rounded to the nearest integer*. This assumption allows the model of speed to have an error $\pm0.5$. It's very unlikely that the values of speed are exactly integers, so it's a reasonable assumption.
- *Breaks are applied at $t = 9$*. The table below shows differences between consecutive speed values, $\Delta v$, and differences between consecutive differences, $\Delta (\Delta v)$. $\Delta (\Delta v)$ is the lowest when $t = 10$. This means that the highest decrease of acceleration occurred between $t = 9$ and $t = 10$, hence we will assume that the breaks were applied during that time. We will model this as if the breaks were applied at $t = 9$.
- *Braking force starts working instantly*. This assumption makes acceleration not continuous and simplifies the calculations.
- *There is no wind*. This allows us to assume that if the plane is not moving then there is no air resistance.

#+ATTR_LATEX: :mode math :environment bmatrix
|----+----+----------+-------------------|
|  t |  v | \Delta v | \Delta (\Delta v) |
|----+----+----------+-------------------|
|  0 | 96 |          |                   |
|  1 | 89 |       -7 |                   |
|  2 | 82 |       -7 |                 0 |
|  3 | 77 |       -5 |                 2 |
|  4 | 72 |       -5 |                 0 |
|  5 | 68 |       -4 |                 1 |
|  6 | 64 |       -4 |                 0 |
|  7 | 61 |       -3 |                 1 |
|  8 | 58 |       -3 |                 0 |
|  9 | 55 |       -3 |                 0 |
| 10 | 50 |       -5 |                -2 |
| 11 | 46 |       -4 |                 1 |
| 12 | 41 |       -5 |                -1 |
| 13 | 38 |       -3 |                 2 |
| 14 | 34 |       -4 |                -1 |
| 15 | 31 |       -3 |                 1 |
| 16 | 27 |       -4 |                -1 |
| 17 | 24 |       -3 |                 1 |
| 18 | 21 |       -3 |                 0 |
| 19 | 18 |       -3 |                 0 |
| 20 | 16 |       -2 |                 1 |
| 21 | 13 |       -3 |                -1 |
| 22 | 10 |       -3 |                 0 |
| 23 |  8 |       -2 |                 1 |
| 24 |  5 |       -3 |                -1 |
| 25 |  3 |       -2 |                 1 |
| 26 |  0 |       -3 |                -1 |
|----+----+----------+-------------------|

** Derivation of the initial model

#+CAPTION: Forces diagram
[[./plane_diagram.png]]

The diagram shows the x-axis, direction of velocity, $v$, the air resistance, $f(v)$, and the constant braking force (when breaks are used), $B$.

$$ F = ma \mbox{ (Newton's second law)} $$
$$ F = \begin{cases} - f(v), & 0 \leq t < 9 \\ - f(v) - B, & 9 \leq t \leq 26 \end{cases} $$
$$ a = \dot{v} $$
$$ \Rightarrow \begin{cases} m\dot{v} = - f(v), & t < 9 \\ m\dot{v} = - f(v) - B, & t \geq 9 \end{cases} $$
$$ \mbox{where } m = 120 000 $$

* Model 1
** Defining f(v)

As you can see in the table with the differences, speed for $t \in (0, 9)$ clearly isn't decreasing linearly. And you can see that, as speed decreases (in that interval), the rate at which it decreases also decreases. So it's reasonable to suggest that

$$ f(v) \propto v^2 $$
$$ \Rightarrow f(v) = kv^2 $$
$$ \Rightarrow m\dot{v} = \begin{cases} - kv^2, & 0 \leq t < 9 \\ - kv^2 - B, & 9 \leq t \leq 26 \end{cases} $$

To make $v$ continuous I will assume that both models are valid for $t = 9$.

$$ m\dot{v} = \begin{cases} - kv^2, & 0 \leq t \leq 9 \\ - kv^2 - B, & 9 \leq t \leq 26 \end{cases} $$

** General solution
*** 0 \leq t \leq 9

$$ m\dot{v} = -kv^2 $$
$$ \Rightarrow m\frac{\dot{v}}{v^2} = -k $$
$$ \Rightarrow m\int \frac{\dot{v}}{v^2} dt = -k\int dt $$
$$ \Rightarrow m\int v^{-2} dv = -kt + c_1 $$
$$ \Rightarrow m\frac{v^{-1}}{-1} = -kt + c_1 $$
$$ \Rightarrow \frac{-m}{v} = -kt + c_1 $$
$$ \Rightarrow v = \frac{m}{kt - c_1} $$

*** 9 \leq t \leq 26

$$ m\dot{v} = -kv^2 - B $$
$$ m\frac{\dot{v}}{kv^2 + B} = -1 $$
$$ m\int \frac{\dot{v}}{kv^2 + B} dt = -\int dt $$
$$ m\int \frac{1}{kv^2 + B} dv = -\int dt $$
$$ \arctan'(\psi) = \frac{1}{\psi^2 + 1} $$
$$ \Rightarrow \arctan'(\frac{a}{b}\psi) = \frac{a}{b} \times \frac{1}{(\frac{a}{b}\psi)^2 + 1} = \frac{a}{b} \times \frac{b^2}{a^2\psi^2 + b^2} = \frac{ab}{a^2\psi^2 + b^2} $$
$$ \Rightarrow \arctan'(\sqrt{\frac{a}{b}}\psi) = \frac{\sqrt{ab}}{a\psi^2 + b} $$
$$ \Rightarrow \int \frac{1}{kv^2 + B} dv = \frac{1}{\sqrt{kB}}\int \frac{\sqrt{kB}}{kv^2 + B} = \frac{\arctan(\sqrt{\frac{k}{B}}v)}{\sqrt{kB}} + c $$
$$ \Rightarrow \frac{m\arctan(\sqrt{\frac{k}{B}}v)}{\sqrt{kB}} = -t + c $$
$$ \Rightarrow \sqrt{\frac{k}{B}}v = \tan \frac{\sqrt{kB}(-t + c)}{m} $$
$$ \Rightarrow v = \sqrt{\frac{B}{k}} \tan \frac{\sqrt{kB}(-t + c)}{m} $$
$$ \mbox{Let } c_2 = \sqrt{kB}c $$
$$ \Rightarrow v = \sqrt{\frac{B}{k}} \tan(\frac{c_2 -\sqrt{kB}t}{m})$$
$$ \Rightarrow v = \begin{cases} \frac{m}{kt - c_1}, & 0 \leq t \leq 9 \\ \sqrt{\frac{B}{k}} \tan(\frac{c_2 -\sqrt{kB}t}{m}), & 9 \leq t \leq 26 \end{cases} $$

** Particular solution

The general solution has 4 unknown constants: the constant braking force, $B$, the coefficient from the air resistance model, $k$, and 2 constants of integration from solving the 2 differential equations, $c_1$ and $c_2$. To make $v$ continuous I will use point $(t = 9, v = 55)$.

$$ \begin{cases} \frac{m}{9k - c_1} = 55 \\ \sqrt{\frac{B}{k}} \tan \frac{c_2 - 9\sqrt{kB}}{m} = 55 \end{cases} $$
$$ \Rightarrow 9k - c_1 = \frac{m}{55} $$
$$ \Rightarrow k = \frac{c_1 + \frac{m}{55}}{9} $$
$$ S \mbox{ is the set of given data.} $$
$$ (t_1, v_1), (t_2, v_2) \in S; t_1, t_2 \leq 9, t_1 \neq t_2 $$
$$ \begin{cases} v_1 = \frac{m}{t_1k - c_1} \\ v_2 = \frac{m}{t_2k - c_1} \end{cases} $$
$$ \Rightarrow \begin{cases} v_1v_2t_2t_1k - v_1v_2t_2c_1 = mv_2t_2 \\ v_1v_2t_1t_2k  - v_1v_2t_1c_1 = mv_1t_1 \end{cases} $$
$$ \Rightarrow c_1v_1v_2(t_2 - t_1) = m(v_1t_1 - v_2t_2) $$
$$ \Rightarrow c_1 = \frac{m(v_1t_1 - v_2t_2)}{v_1v_2(t_2 - t_1)} $$
$$ \sqrt{\frac{B}{k}} \tan \frac{c_2 - 9\sqrt{kB}}{m} = 55 $$
$$ \Rightarrow c_2(B) = m\arctan(\frac{55k\sqrt{\frac{B}{k}}}{B}) + 9\sqrt{Bk} $$
$$ (t_3, v_3) \in S; t_3 > 9 $$
$$ \Rightarrow v_3 = \sqrt{\frac{B}{k}} \tan \frac{c_2(B) - \sqrt{kB}t_3}{m} $$
$$ \Rightarrow \arctan(v_3\sqrt{\frac{k}{B}}) = \frac{c_2(B) - \sqrt{kB}t_3}{m} $$
$$ \mbox{Let } \lambda(B) = \frac{c_2(B) - \sqrt{kB}t_3}{m} - \arctan(v_3\sqrt{\frac{k}{B}}) $$
$$ \mbox{Newton-Raphson's rule: } B_{n+1} = B_n + \frac{\lambda'(B_n)}{\lambda(B_n)} $$
\begin{equation}
  \begin{cases}
    c_1 = \frac{m(v_1t_1 - v_2t_2)}{v_1v_2(t_2 - t_1)} \\
    k = \frac{c_1 + \frac{m}{55}}{9} \\
    c_2(B) = m\arctan(\frac{55k\sqrt{\frac{B}{k}}}{B}) + 9\sqrt{Bk} \\
    \lambda(B) = \frac{c_2(B) - \sqrt{kB}t_3}{m} - \arctan(v_3\sqrt{\frac{k}{B}}) \\
    B_{n+1} = B_n - \frac{\lambda(B_n)}{\lambda'(B_n)}
    B_0 = 300000
  \end{cases}
\end{equation}

To find $B_0$ I used a graphing web app. First I put in all the formulas and given data points. $v(0) = \frac{m}{- c_1} > 0$, hence $c_1 < 0$. I started decreasing $c_1$ until it roughly matched first 10 points. $B > 0$, so for this value of $c_1$ I started increasing $B$ until I found that $B \approx 300,000$.

[[./img/not_optimized_first_model.png]]

Now I can use programming to find the parameters. To do this I wrote 5 modules in Haskell. The first module is a module for automatic differentiation using dual numbers and Newton-Raphson method.

#+INCLUDE: "./optimizer/src/AutoDiff.hs" src haskell

The second module contains the given data.

#+INCLUDE: "./optimizer/src/GivenData.hs" src haskell

Then I defined a general interface for models.

#+INCLUDE: "./optimizer/src/Solution/Model.hs" src haskell

The fourth module is for the first model and finding parameters.

#+INCLUDE: "./optimizer/src/Solution/First.hs" src haskell

And finally the main module, which runs this code to find the parameters.

#+BEGIN_SRC haskell
import           Data.List      (sort, sortOn)
import           Data.Proxy

import           GivenData
import qualified Solution.First as Sol
import           Solution.Model as Model

adjustLength :: (a -> String) -> (a -> String -> b) -> [a] -> [b]
adjustLength g s l =
  (\x -> let str = g x in s x $ str ++ replicate (longest - length str) ' ') <$> l
  where longest = maximum $ length . g <$> l

showPredictions :: (ModelParams f, Show (f Double)) => f Double -> String
showPredictions params =
  foldMap (\(x, y, z, w) -> "| " ++ x ++ " | " ++ y ++ " | " ++ z ++ " | " ++ w ++ " |\n")
     $  adjustLength (\(_, _, _, w) -> w) (\(x, y, z, _) w -> (x, y, z, w))
     $  adjustLength (\(_, _, z, _) -> z) (\(x, y, _, w) z -> (x, y, z, w))
     $  adjustLength (\(_, y, _, _) -> y) (\(x, _, z, w) y -> (x, y, z, w))
     $  adjustLength (\(x, _, _, _) -> x) (\(_, y, z, w) x -> (x, y, z, w))
     $  (("Time", "Predicted Speed", "Rounded predicted speed", "Observed Speed"):)
     $  (\(Point t v) -> let v' = predict params t in (show t, show v', rounded v', rounded v))
    <$> points
  where rounded = show . round

maxErr :: (a -> Double -> Double) -> a -> Double
maxErr predict params =
  maximum $ abs . (\(Point t v) -> predict params t - v) <$> points
  where n = fromIntegral (length points)

avrErr :: (a -> Double -> Double) -> a -> Double
avrErr predict params =
  (/ n) $ sum $ abs . (\(Point t v) -> predict params t - v) <$> points
  where n = fromIntegral (length points)

rms :: (a -> Double -> Double) -> a -> Double
rms predict params =
  (/ n) $ sqrt $ sum $ (^2) . (\(Point t v) -> predict params t - v) <$> points
  where n = fromIntegral (length points)

showErrors :: (ModelParams f, Show (f Double)) => f Double -> String
showErrors params =
  foldMap (\(x, y, z) -> "| " ++ x ++ " | " ++ y ++ " | " ++ z ++ " |\n")
  $  adjustLength (\(_, _, z) -> z) (\(x, y, _) z -> (x, y, z))
  $  adjustLength (\(_, y, _) -> y) (\(x, _, z) y -> (x, y, z))
  $  adjustLength (\(x, _, _) -> x) (\(_, y, z) x -> (x, y, z))
  [ ("Maximum Error", "Average Error", "Root mean square")
  , (show $ maxErr predict params, show $ avrErr predict params, show $ rms predict params)
  ]

showResults :: (ModelParams f, Show (f Double)) => f Double -> String
showResults params = show params ++ "\n"
                  ++ showErrors params ++ "\n"
                  ++ showPredictions params

results :: [Sol.Parameters Double]
results =  sortOn (rms predict) $ findParams
      <$> combinations (Proxy :: Proxy Sol.Parameters)

main :: IO ()
main = do putStrLn "The best result:"
          putStrLn $ showResults $ head results
          putStrLn "The worst result:"
          putStr   $ showResults $ last results
#+END_SRC

Here is the result. It shows the best and the worst parameters found using the derived formulas and for both of the configurations it shows sums of squares of errors, and predictions for each of the given point with rounded versions.

#+NAME: first_model_code_result
[[./img/first_model_code_result.png]]

** Graphs and Analysis

|------+--------------------+-------------------------+----------------|
| Time |    Predicted Speed | Rounded predicted speed | Observed Speed |
|------+--------------------+-------------------------+----------------|
|  0.0 |               96.0 |                      96 |             96 |
|------+--------------------+-------------------------+----------------|
|  1.0 |  88.65671641791045 |                      89 |             89 |
|------+--------------------+-------------------------+----------------|
|  2.0 |  82.35701906412478 |                      82 |             82 |
|------+--------------------+-------------------------+----------------|
|  3.0 |  76.89320388349515 |                      77 |             77 |
|------+--------------------+-------------------------+----------------|
|  4.0 |   72.1092564491654 |                      72 |             72 |
|------+--------------------+-------------------------+----------------|
|  5.0 |  67.88571428571429 |                      68 |             68 |
|------+--------------------+-------------------------+----------------|
|  6.0 |  64.12955465587045 |                      64 |             64 |
|------+--------------------+-------------------------+----------------|
|  7.0 | 60.767263427109974 |                      61 |             61 |
|------+--------------------+-------------------------+----------------|
|  8.0 | 57.739975698663415 |                      58 |             58 |
|------+--------------------+-------------------------+----------------|
|  9.0 |  54.99999999999999 |                      55 |             55 |
|------+--------------------+-------------------------+----------------|
| 10.0 |  50.10816658852111 |                      50 |             50 |
|------+--------------------+-------------------------+----------------|
| 11.0 | 45.622063036212374 |                      46 |             46 |
|------+--------------------+-------------------------+----------------|
| 12.0 | 41.475988862696376 |                      41 |             41 |
|------+--------------------+-------------------------+----------------|
| 13.0 |  37.61659930730753 |                      38 |             38 |
|------+--------------------+-------------------------+----------------|
| 14.0 |  34.00002062308155 |                      34 |             34 |
|------+--------------------+-------------------------+----------------|
| 15.0 | 30.589724513917773 |                      31 |             31 |
|------+--------------------+-------------------------+----------------|
| 16.0 |  27.35493651754378 |                      27 |             27 |
|------+--------------------+-------------------------+----------------|
| 17.0 | 24.269426053394415 |                      24 |             24 |
|------+--------------------+-------------------------+----------------|
| 18.0 |   21.3105731783461 |                      21 |             21 |
|------+--------------------+-------------------------+----------------|
| 19.0 |  18.45863841578133 |                      18 |             18 |
|------+--------------------+-------------------------+----------------|
| 20.0 | 15.696183121094782 |                      16 |             16 |
|------+--------------------+-------------------------+----------------|
| 21.0 |  13.00760227741885 |                      13 |             13 |
|------+--------------------+-------------------------+----------------|
| 22.0 | 10.378741615250261 |                      10 |             10 |
|------+--------------------+-------------------------+----------------|
| 23.0 |  7.796577949697343 |                       8 |              8 |
|------+--------------------+-------------------------+----------------|
| 24.0 |  5.248946558956161 |                       5 |              5 |
|------+--------------------+-------------------------+----------------|
| 25.0 | 2.7243028986249413 |                       3 |              3 |
|------+--------------------+-------------------------+----------------|
| 26.0 | 0.2115083630415522 |                       0 |              0 |
|------+--------------------+-------------------------+----------------|

|---------------------+---------------------+----------------------|
|       Maximum Error |       Average Error |     Root mean square |
|---------------------+---------------------+----------------------|
| 0.47598886269637575 | 0.23822424753496707 | 5.344098631088544e-2 |
|---------------------+---------------------+----------------------|

I'll now look at the two configurations that I obtained and plot them. The red line shows the worst configuration, and the purple line shows the best configuration.

[[./img/first_model_graph_result.png]]

As you can see in [[first_model_code_result]] rounded predictions of the best configuration exactly match the given data.

** Conclusion of Model 1

\begin{equation}
  \begin{cases}
    v(t) = \frac{m}{kt - c_1}, & 0 \leq t \leq 9 \\
    v(t) = \sqrt{\frac{B}{k}} \tan(\frac{c_2 -\sqrt{kB}t}{m}), & 9 \leq t \leq 26 \\
    c_1 = -1250.0 \\
    k = 103.\dot{5}\dot{3} \\
    c_2 = 145677.3177225051 \\
    B = 301257.94278185006 \\
  \end{cases}
\end{equation}

$$ x = \int_0^{26} v dt = \int_0^9 v dt + \int_9^{26} v dt $$
$$ \int \frac{m}{kt - c_1} dt = \frac{m}{k} \int \frac{k}{kt - c_1} dt = \frac{m}{k}\ln(kt - c_1) + \mbox{constant} $$
$$ \int \sqrt{\frac{B}{k}}\tan(\frac{c_2 -\sqrt{kB}t}{m}) dt = \sqrt{\frac{B}{k}} \times \frac{-m}{\sqrt{kB}} \int \tan(\frac{c_2 - \sqrt{kB}t}{m})\times\frac{-\sqrt{kB}}{m} dt =$$
$$ = \frac{m}{k} \int \frac{-\sin(u)}{\cos{u}} du = \frac{m}{k} \ln(\cos(\frac{c_2 - \sqrt{kB}t}{m})) + \mbox{constant} $$
$$ \Rightarrow x = \left[\frac{m}{k}\ln(kt - c_1)\right]_0^9 + \left[\frac{m}{k} \ln(\cos(\frac{c_2 - \sqrt{kB}t}{m}))\right]_9^{26} = \frac{m}{k}\left(\ln(\frac{c_1 - 9k}{c_1}) + \ln(\frac{\cos(\frac{c_2 - 26\sqrt{kB}}{m})}{\cos(\frac{c_2 - 9\sqrt{kB}}{m})})\right) $$
$$ \Rightarrow x = \frac{m}{k}\ln\left(\frac{(c_1 - 9k)\cos(\frac{c_2 - 26\sqrt{kB}}{m})}{c_1\cos(\frac{c_2 - 9\sqrt{kB}}{m})}\right) $$

[[./img/first_model_integral.png]]

$$ \Rightarrow x \approx 1058.650842313513 \approx 1060 $$

The aeroplane fully stops in approximately $1060$ meters after touchdown, hence the runway has to be at least $1060$ meters long. However, I would advise length of $2000$ to be safe even if conditions change, for example in case touchdown occurs further, there is wind, friction between the wheels and the runway is lower.

* Model 2
** Redefining f(v)
** General solution
*** 0 \leq t < 9

$$ m\dot{v} = -kv^2 - \lambda v $$
$$ \Rightarrow m\int \frac{1}{kv^2 + \lambda v} dv = -t + c_3 $$
$$ kv^2 + \lambda v \equiv v(kv + \lambda)  $$
$$ \frac{1}{v(kv + \lambda)} \equiv \frac{A}{v} + \frac{B}{kv + \lambda} $$
$$ \Rightarrow 1 \equiv A(kv + \lambda) + Bv $$
$$ \Rightarrow \begin{cases} A = \frac{1}{\lambda} \\ B = -\frac{k}{\lambda} \end{cases} $$
$$ \Rightarrow \frac{1}{kv^2 + \lambda v} \equiv \frac{1}{\lambda v} - \frac{k}{k \lambda v + \lambda^2} ( \equiv \frac{k \lambda v + \lambda^2 - k \lambda v}{\lambda v (k \lambda v + \lambda^2)} \equiv \frac{\lambda^2}{\lambda^2(kv^2 + \lambda v)} \equiv \frac{1}{kv^2 + \lambda v} ) $$
$$ \Rightarrow \int \frac{1}{kv^2 - \lambda v} dv = \frac{1}{\lambda}\int v^{-1} dv - \frac{1}{\lambda}\int \frac{k}{kv + \lambda} dv $$
$$ \Rightarrow \frac{m}{\lambda}(\ln(v) - \ln(kv + \lambda)) = c_3 - t $$
$$ \Rightarrow \ln(\frac{v}{kv + \lambda}) = \frac{\lambda}{m} (c_3 - t) $$
$$ \Rightarrow \frac{v}{kv + \lambda} = e^{\frac{\lambda}{m} (c_3 - t)} $$
$$ \Rightarrow v = k e^{\frac{\lambda}{m} (c_3 - t)} v + \lambda e^{\frac{\lambda}{m} (c_3 - t)} $$
$$ v = \frac{\lambda e^{\frac{\lambda}{m} (c_3 - t)}}{1 - k e^{\frac{\lambda}{m} (c_3 - t)}} $$
$$ v = \frac{\lambda e^{\frac{\lambda}{m} c_3}}{e^{\frac{\lambda}{m} t} - ke^{\frac{\lambda}{m}c_3}} $$

*** 9 < t \leq 26

$$ m\dot{v} = -kv^2 - \lambda v - W $$
$$ \Rightarrow m\int \frac{1}{kv^2 + \lambda v + W} dv = -\int dt $$
$$ kv^2 + \lambda v + W = (\sqrt{k}v + \frac{\lambda}{2\sqrt{k}})^2 + W - \frac{\lambda^2}{4k} = (W - \frac{\lambda^2}{4k})((\frac{\sqrt{k}v + \frac{\lambda}{2\sqrt{k}}}{\sqrt{W - \frac{\lambda^2}{4k}}})^2 + 1) = (W - \frac{\lambda^2}{4k})((\frac{2kv + \lambda}{\sqrt{4kW - \lambda^2}})^2 + 1) $$
$$ \mbox{Let } h(x) = \frac{2kx + \lambda}{\sqrt{4kW - \lambda^2}} $$
$$ \Rightarrow h'(x) = \frac{2k}{\sqrt{4kW - \lambda^2}} $$
$$ \arctan'(x) = \frac{1}{1 + x^2} $$
$$ \Rightarrow (\arctan(h(x)))' = \arctan'(h(x))h'(x) = \frac{1}{h^2(x) + 1}h'(x) = \frac{\frac{2k}{\sqrt{4kW - \lambda^2}}}{(\frac{2kx + \lambda}{\sqrt{4kW - \lambda^2}})^2 + 1} $$
$$ \Rightarrow m\int \frac{1}{kv^2 + \lambda v + W} dv = m\int \frac{1}{(W - \frac{\lambda^2}{4k})((\frac{2kv + \lambda}{\sqrt{4kW - \lambda^2}})^2 + 1)} dv = \frac{2m\sqrt{4kW - \lambda^2}}{4k(W - \frac{\lambda^2}{4k})}\int \frac{\frac{2k}{\sqrt{4kW - \lambda^2}}}{(\frac{2kv + \lambda}{\sqrt{4kW - \lambda^2}})^2 + 1} dv $$
$$ \Rightarrow \frac{2m}{\sqrt{4kW-\lambda^2}}\arctan(\frac{2kv + \lambda}{\sqrt{4kW - \lambda^2}}) = -t + c_4 $$
$$ \Rightarrow \frac{2kv + \lambda}{\sqrt{4kW - \lambda^2}} = \tan(\frac{\sqrt{4kW-\lambda^2}}{2m}(c_4 - t)) $$
$$ \Rightarrow v = \frac{\sqrt{4kW - \lambda^2}\tan(\frac{\sqrt{4kW-\lambda^2}}{2m}(c_4 - t)) - \lambda}{2k}  $$
$$ \Rightarrow v = \begin{cases}\frac{\lambda e^{\frac{\lambda}{m} c_3}}{e^{\frac{\lambda}{m} t} - ke^{\frac{\lambda}{m}c_3}}, & t \in [0, 9] \\ \frac{\sqrt{4kW - \lambda^2}\tan(\frac{\sqrt{4kW-\lambda^2}}{2m}(c_4 - t)) - \lambda}{2k}, & t \in [9, 26] \end{cases} $$

** Particular solution

$$ \Rightarrow \begin{cases}\frac{\lambda e^{\frac{\lambda}{m} c_3}}{e^{9\frac{\lambda}{m}} - ke^{\frac{\lambda}{m}c_3}} = 55 \\ \frac{\sqrt{4kW - \lambda^2}\tan(\frac{\sqrt{4kW-\lambda^2}}{2m}(c_4 - 9)) - \lambda}{2k} = 55 \end{cases} $$
$$ \mbox{Let } (t_1, v_1), (t_2, v_2) \in S; t_1, t_2 \in [0, 9]; t_1 > t_2 $$
$$ \frac{\lambda e^{\frac{\lambda}{m} c_3}}{e^{t_1\frac{\lambda}{m}} - ke^{\frac{\lambda}{m}c_3}} = v_1 $$
$$ \Rightarrow v_1ke^{\frac{\lambda}{m}c_3} = v_1e^{t_1\frac{\lambda}{m}} - \lambda e^{\frac{\lambda}{m} c_3} $$
$$ \Rightarrow k = \frac{v_1e^{t_1\frac{\lambda}{m}} - \lambda e^{\frac{\lambda}{m} c_3}}{v_1e^{\frac{\lambda}{m}c_3}} $$
$$ \Rightarrow k = e^{\frac{\lambda}{m}(t_1 - c_3)} - \frac{\lambda}{v_1} $$
$$ \frac{m}{\lambda}(\ln(v_2) - \ln(kv_2 + \lambda)) = c_3 - t_2 $$
$$ \Rightarrow c_3 = \frac{m}{\lambda}(\ln(v_2) - \ln(kv_2 + \lambda)) + t_2 $$
$$ \Rightarrow c_3 = \frac{m}{\lambda}(\ln(\frac{v_2}{v_2 (e^{\frac{\lambda}{m}(t_1 - c_3)} - \frac{\lambda}{v_1}) + \lambda})) + t_2 $$
$$ \Rightarrow e^{\frac{\lambda (c_3 - t_2)}{m}} = \frac{v_2}{v_2 (e^{\frac{\lambda}{m}(t_1 - c_3)} - \frac{\lambda}{v_1}) + \lambda} $$
$$ \Rightarrow e^{\frac{\lambda}{m}c_3}e^{-\frac{\lambda}{m}t_2} = \frac{v_2}{v_2 (e^{\frac{\lambda}{m}t_1}e^{-\frac{\lambda}{m}c_3} - \frac{\lambda}{v_1}) + \lambda} $$
$$ \Rightarrow (v_2 e^{\frac{\lambda}{m}t_1}e^{-\frac{\lambda}{m}c_3} - v_2\frac{\lambda}{v_1} + \lambda)e^{\frac{\lambda}{m}c_3}e^{-\frac{\lambda}{m}t_2} = v_2 $$
$$ \Rightarrow v_2 e^{\frac{\lambda}{m}(t_1 - t_2)} - (v_2\frac{\lambda}{v_1} - \lambda)e^{\frac{\lambda}{m}(c_3 - t_2)} = v_2 $$
$$ \Rightarrow e^{\frac{\lambda}{m}(c_3 - t_2)} = \frac{v_1v_2}{\lambda(v_2 - v_1)} (e^{\frac{\lambda}{m}(t_1 - t_2)} - 1) $$
$$ \Rightarrow e^{\frac{\lambda}{m}c_3} = \frac{v_1v_2}{\lambda(v_2 - v_1)} (e^{\frac{\lambda}{m}t_1} - e^{\frac{\lambda}{m}t_2}) $$
$$ \Rightarrow c_3 = \frac{m}{\lambda}\ln(\frac{v_1v_2(e^{\frac{\lambda}{m}t_1} - e^{\frac{\lambda}{m}t_2})}{\lambda(v_2 - v_1)}) $$

optimizer/ChangeLog.md
# Changelog for optimizer

## Unreleased changes


optimizer.hs
{-# LANGUAGE GeneralizedNewtypeDeriving #-}

import           Data.List (sortOn)

vs :: [Integer]
vs = [96,89,82,77,72,68,64,61,58,55,50,46,41,38,34,31,27,24,21,18,16,13,10,8,5,3,0]

data GivenPoint =
    GivenPoint { gpTime  :: Integer
               , gpSpeed :: Integer
               } deriving (Show, Eq)

points :: [GivenPoint]
points = zipWith GivenPoint [0..26] vs

newtype ResFactor = ResFactor { unResFactor :: Double } deriving (Show, Eq, Num, Ord)
newtype IntConst  = IntConst  { unIntConst  :: Double } deriving (Show, Eq, Num, Ord)

mass :: Integer
mass = 12*10^4

predictSpeed :: ResFactor -> IntConst -> Integer -> Double
predictSpeed rf ic time
  | 0 <= time && time < 9 = fromInteger mass / (unResFactor rf * fromInteger time - unIntConst ic)
  | otherwise = 0

loss :: ResFactor -> IntConst -> Double
loss rf ic = sum $ (^2) <$> zipWith (-) (fromInteger . gpSpeed <$> points) (predictSpeed rf ic . gpTime <$> points)

getResFactor :: GivenPoint -> GivenPoint -> ResFactor
getResFactor (GivenPoint t1 v1) (GivenPoint t2 v2) =
    ResFactor $ fromInteger (mass * (v2 - v1)) / fromInteger (v1 * v2 * (t1 - t2))

getIntConst :: GivenPoint -> GivenPoint -> IntConst
getIntConst (GivenPoint t1 v1) (GivenPoint t2 v2) =
    IntConst $ fromInteger (mass * (t2*v2 - t1*v1)) / fromInteger (v1 * v2 * (t1 - t2))

main :: IO ()
main =  print $  head $ sortOn (\(_, _, l) -> l)
     $  (\(rf, ic) -> (rf, ic, loss rf ic)) . (\(p1, p2) -> (getResFactor p1 p2, getIntConst p1 p2))
    <$> filter (uncurry (/=)) [ (p1, p2) | p1 <- somePoints, p2 <- somePoints ]
     where somePoints = filter (\p -> gpTime p < 9) points
